packages/teleport/src/DesktopSession/DesktopSession.tsx

```diff
@@ -40,42 +46,117 @@ export function DesktopSession(props: State) {
-  } else if (clipboardError) {
-    alertText = clipboardState.errorText || 'clipboard sharing failed';
-  } else if (unknownConnectionError) {
-    alertText = 'Session disconnected for an unknown reason';
-  }
+  // Manages the state of the error dialog.
+  const [errorDialog, setErrorDialog] = useState({
+    open: false,
+    text: '',
+    fatal: false,
+  });
+
+  // onDialogClose is called when a user
+  // dismisses a non-fatal error dialog.
+  const onDialogClose = () => {
+    // This setTdpConnection call will cause the useEffect below
+    // to calculate the errorDialog state.
+    setTdpConnection(prevState => {
+      // onDialogClose should only be called when
+      // the user dismisses a non-fatal error dialog,
+      // and prevState.status === '' means non-fatal
+      // error dialog, so the below if statement should
+      // always be true.
+      if (prevState.status === '') {
+        // If prevState.status was a non-fatal error,
+        // we assume that the TDP connection remains open.
+        return { status: 'success' };
+      }
+      return prevState;
+    });
+  };
 
-  if (alertText) {
+  // Calculate the state of the error dialog based on the various
+  // sub-states which determine it.
+  useEffect(() => {
+    const clipboardError = clipboardState.enabled && clipboardState.errorText;
+
+    // Websocket is closed but we haven't
+    // closed it on purpose or registered a fatal tdp error.
+    const unknownConnectionError =
+      wsConnection === 'closed' &&
+      !disconnected &&
+      (tdpConnection.status === 'success' || tdpConnection.status === '');
+
+    let errorText = '';
+    if (fetchAttempt.status === 'failed') {
+      errorText = fetchAttempt.statusText || 'fetch attempt failed';
+    } else if (tdpConnection.status === 'failed') {
+      errorText = tdpConnection.statusText || 'tdp connection failed';
+    } else if (tdpConnection.status === '') {
+      errorText = tdpConnection.statusText || 'encountered a non-fatal error';
+    } else if (clipboardError) {
+      errorText = clipboardState.errorText || 'clipboard sharing failed';
+    } else if (unknownConnectionError) {
+      errorText = 'Session disconnected for an unknown reason';
+    }
+
+    const open = errorText !== '';
+    const fatal = tdpConnection.status !== '';
+
+    setErrorDialog({ open, text: errorText, fatal });
+  }, [clipboardState, fetchAttempt, tdpConnection, wsConnection, disconnected]);
+
+  if (errorDialog.open) {
     return (
       <Session {...props}>
-        <DesktopSessionAlert my={2} mx={10} children={alertText} />
+        <Dialog
+          dialogCss={() => ({ width: '484px' })}
+          onClose={onDialogClose}
+          open={errorDialog.open}
+        >
+          <DialogHeader style={{ flexDirection: 'column' }}>
+            {errorDialog.fatal && <DialogTitle>Fatal Error</DialogTitle>}
+            {!errorDialog.fatal && (
+              <DialogTitle>Dismiss to Continue</DialogTitle>
```

Comment:
I think it will be better to say something like `Unsupported Action` or something among those lines so at least i get the big picture before i read the message